### PR TITLE
cdc/codec: Fix canal json encoding

### PR DESCRIPTION
**Pull request by 3AceShowHand**
### What problem does this PR solve? 

This PR is a part of make Canal-JSON GA.

For Canal-JSON protocol, we should make sure TiCDC generates output almost identical to the official implementation, this PR tries to make it happen. 

### What is changed and how it works?

* Fix TiCDC encoding Canal-JSON process, especially for `MySQLType`, `SQLType`, data value format.
* add more unit test to cover all type mapping and value encoding and decoding process.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

Side effects

Related changes
### Release note 

```release-note
Fix canal-json protocol to be more compatible with the official.
```

### CODE REVIEW

File: cdc/sink/codec/canal_flat_test.go [hunk 1]
```diff
@@ -337,6 +337,18 @@ func (s *canalFlatSuite) TestEncodeCheckpointEvent(c *check.C) {
...
+//	CommitTs: 429457251303161859,
+//	Table: &model.TableName{
+//		Schema: "cdc_tiflash_test",
+//		Table:  "multi_data_type",
+//	},
+//	Columns: []*model.Column{
+//		{Name: "id"},
+//	},
+//	PreColumns: nil,
+//}
```
Comment (overvenus) [inline cdc/sink/codec/canal_flat_test.go:350]: Could you uncomment or remove?

File: cdc/sink/codec/java.go [hunk 2]
```diff
@@ -63,13 +64,12 @@ const (
 	// JavaSQLTypeTIMESTAMP_WITH_TIMEZONE JavaSQLType = 2014
 )
 
-// MysqlToJavaType converts the mysql protocol types to java sql types
-func MysqlToJavaType(mysqlType byte) JavaSQLType {
+// MySQLType2JavaType converts the mysql protocol types to java sql types
+// take https://github.com/alibaba/canal/blob/master/dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/event/RowsLogBuffer.java#L132 as reference
```
Comment (overvenus) [inline cdc/sink/codec/java.go:68]:
```suggestion
// take https://github.com/alibaba/canal/blob/b54bea5e3/dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/event/RowsLogBuffer.java#L1 as reference
```

Please use permalink.

File: cdc/sink/codec/canal.go [hunk 3]
```diff
@@ -20,13 +20,14 @@ import (
 	"strconv"
 	"strings"
 
+	"github.com/pingcap/tidb/parser/mysql"
+
```
Comment (overvenus) [inline cdc/sink/codec/canal.go:24]:
```suggestion
	"github.com/pingcap/tidb/parser/mysql"
```

File: cdc/sink/codec/canal_flat_test.go (hunk 1 above)
Comment (3AceShowHand) [reply cdc/sink/codec/canal_flat_test.go:350]: The PR is not ready to be reviewed.
this is a very important test case which will be filled in later.

File: cdc/sink/codec/canal_flat.go [hunk 4]
```diff
@@ -261,14 +278,19 @@ func (c *CanalFlatEventBatchEncoder) newFlatMessageForDML(e *model.RowChangedEve
...
-	flatMessage.Old = append(flatMessage.Old, oldData)
+	if e.IsDelete() {
+		flatMessage.Data = append(flatMessage.Data, oldData)
+	} else if e.IsInsert() {
+		flatMessage.Data = append(flatMessage.Data, data)
+	} else if e.IsUpdate() {
+		flatMessage.Old = make([]map[string]interface{}, 0)
+		flatMessage.Old = append(flatMessage.Old, oldData)
+		flatMessage.Data = append(flatMessage.Data, data)
+	}
```
Comment (overvenus) [inline cdc/sink/codec/canal_flat.go:293]: Could `e` be false for `IsDelete`, `IsInsert` and `IsUpdate`? If not, please add an assertion here.

```suggestion
	} else { log.Panic("unreachable xxx") }
```

File: cdc/sink/codec/canal_flat.go [hunk 5]
```diff
@@ -509,33 +531,36 @@ func canalFlatMessage2RowChangedEvent(flatMessage canalFlatMessageInterface) (*m
...
-		tp := types.StrToType(typeStr)
+		mysqlTypeStr, ok := mysqlType[name]
 		if !ok {
+			// this should not happen, else we have to check encoding for mysqlType.
 			return nil, cerrors.ErrCanalDecodeFailed.GenWithStack(
-				"mysql type does not found, column: %+v, type: %+v", name, tp)
+				"mysql type does not found, column: %+v, mysqlType: %+v", name, mysqlType)
 		}
-		col := NewColumn(value, tp).ToSinkColumn(name)
+		mysqlTypeStr = strings.TrimSuffix(mysqlTypeStr, " unsigned")
```
Comment (overvenus) [inline cdc/sink/codec/canal_flat.go:561]:
```suggestion
		mysqlTypeStr = strings.TrimSuffix(mysqlTypeStr, "unsigned")
```

A typo?

Could you add some comments about what `mysqlTypeStr` would look like?

**Review by overvenus — COMMENTED** [review]
Rest LGTM

**Comment by hi-rustin** [comment]
@3AceShowHand Please refine your PR title and body.

File: cdc/sink/codec/canal_flat.go (hunk 5 above)
Comment (3AceShowHand) [reply cdc/sink/codec/canal_flat.go:561]: no. when encoding, we append " unsigned" to unsigned MySQL types, this just reverse it.

File: cdc/sink/codec/json.go [hunk 6]
```diff
@@ -23,6 +23,8 @@ import (
 	"strconv"
 	"strings"
 
+	"golang.org/x/text/encoding/charmap"
+
```
Comment (hi-rustin) [inline cdc/sink/codec/json.go:27]:
```suggestion
```

**Review by hi-rustin — COMMENTED** [review]
Looks good! I'll spend some more time tomorrow to see if the type correspondence is correct.

File: cdc/sink/codec/canal.go [hunk 7]
```diff
@@ -133,84 +134,163 @@ func (b *canalEntryBuilder) buildHeader(commitTs uint64, schema string, table st
...
+	// no need to promote type for `year` and `bit`
+	if c.Type == mysql.TypeYear || c.Type == mysql.TypeBit {
+		return javaType
+	}
+
+	if c.Type == mysql.TypeFloat || c.Type == mysql.TypeDouble || c.Type == mysql.TypeNewDecimal {
+		return javaType
+	}
+
+	// for **unsigned** `Int` related data type, should have type in `uint64`. see reference:
```
Comment (liuzix) [inline cdc/sink/codec/canal.go:163]: Do you mean "integral types"?

File: cdc/sink/codec/canal_flat.go [hunk 8]
```diff
@@ -261,14 +278,21 @@ func (c *CanalFlatEventBatchEncoder) newFlatMessageForDML(e *model.RowChangedEve
...
-	// even if the element could be nil. Changing this could break Alibaba's adapter
-	flatMessage.Data = append(flatMessage.Data, data)
-	flatMessage.Old = append(flatMessage.Old, oldData)
+	if e.IsDelete() {
+		flatMessage.Data = append(flatMessage.Data, oldData)
+	} else if e.IsInsert() {
+		flatMessage.Data = append(flatMessage.Data, data)
+	} else if e.IsUpdate() {
+		flatMessage.Old = make([]map[string]interface{}, 0)
+		flatMessage.Old = append(flatMessage.Old, oldData)
```
Comment (liuzix) [inline cdc/sink/codec/canal_flat.go:291]: Can you combine the lines and use a map literal? It would be more readable this way.

**Review by liuzix — APPROVED** [review]

File: cdc/sink/codec/java.go [hunk 9]
```diff
@@ -146,6 +141,10 @@ func MysqlToJavaType(mysqlType byte) JavaSQLType {
 	case mysql.TypeBit:
 		return JavaSQLTypeBIT
 
+	case mysql.TypeJSON:
+		// json: see jdbc 8.0, https://github.com/mysql/mysql-connector-j/blob/8.0.20/src/main/core-api/java/com/mysql/cj/MysqlType.java
```
Comment (overvenus) [inline cdc/sink/codec/java.go:145]: Please use permalink https://github.com/mysql/mysql-connector-j/blob/afc0a1/src/main/core-api/java/com/mysql/cj/MysqlType.java#L2

File: cdc/sink/codec/canal.go [hunk 10]
```diff
@@ -133,84 +134,163 @@ func (b *canalEntryBuilder) buildHeader(commitTs uint64, schema string, table st
...
+
+	if types.IsTypeChar(c.Type) {
+		return strings.Replace(mysqlType, "char", "binary", 1)
+	}
+
+	return mysqlType
+}
+
+// build the Column in the canal RowData
+// reference: https://github.com/alibaba/canal/blob/master/parse/src/main/java/com/alibaba/otter/canal/parse/inbound/mysql/dbsync/LogEventConvert.java#L756-L872
```
Comment (overvenus) [inline cdc/sink/codec/canal.go:278]: Please use permalink.

File: cdc/sink/codec/canal_flat.go [hunk 11]
```diff
@@ -82,6 +82,7 @@ type canalFlatMessageInterface interface {
 	getOld() map[string]interface{}
 	getData() map[string]interface{}
 	getMySQLType() map[string]string
+	getJavaSQLType() map[string]int32
 }
 
 // adapted from https://github.com/alibaba/canal/blob/master/protocol/src/main/java/com/alibaba/otter/canal/protocol/FlatMessage.java
```
Comment (overvenus) [inline cdc/sink/codec/canal_flat.go:88]: Please use permalink.

**Review by overvenus — APPROVED** [review]

File: cdc/sink/codec/java.go [hunk 12]
```diff
@@ -63,13 +63,12 @@ const (
 	// JavaSQLTypeTIMESTAMP_WITH_TIMEZONE JavaSQLType = 2014
 )
 
-// MysqlToJavaType converts the mysql protocol types to java sql types
-func MysqlToJavaType(mysqlType byte) JavaSQLType {
+// mySQLType2JavaType converts the mysql protocol types to java sql types
+// take https://github.com/alibaba/canal/blob/b54bea5e3/dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/event/RowsLogBuffer.java#L1 as reference
```
Comment (hi-rustin) [inline cdc/sink/codec/java.go:67]:
```suggestion
// take https://github.com/alibaba/canal/blob/b54bea5e3/dbsync/src/main/java/com/taobao/tddl/dbsync/binlog/event/RowsLogBuffer.java#L132 as reference
```

File: cdc/sink/codec/java.go [hunk 13]
```diff
@@ -39,6 +39,7 @@ const (
 	JavaSQLTypeLONGVARBINARY JavaSQLType = -4
 	JavaSQLTypeNULL          JavaSQLType = 0
 	JavaSQLTypeBLOB          JavaSQLType = 2004
+	JavaSQLTypeCLOB          JavaSQLType = 2005
```
Comment (hi-rustin) [inline cdc/sink/codec/java.go:43]: JavaSQLTypeLONGVARCHAR doesn't seem to be used?

**Review by hi-rustin — APPROVED** [review]

**Comment by nkorange** [comment]
Hi, may I know if this PR is applicable to 5.1.4?

**Comment by hi-rustin** [comment]
> Hi, may I know if this PR is applicable to 5.1.4?

It was not carried over to the old version.

**Pull request merged**